src/entrypoint/graphql-admin-server.ts

```diff
@@ -0,0 +1,187 @@
+import { ApolloServer } from "apollo-server-express"
+import dotenv from "dotenv"
+import express from "express"
+import expressJwt from "express-jwt"
+import { applyMiddleware } from "graphql-middleware"
+import { and, rule, shield } from "graphql-shield"
+import mongoose from "mongoose"
+import pino from "pino"
+import PinoHttp from "pino-http"
+import { v4 as uuidv4 } from "uuid"
+
+import { yamlConfig } from "../config"
+import { AuthorizationError, IPBlacklistedError } from "../error"
+import { baseLogger } from "../logger"
+import { redis } from "../redis"
+import { User } from "../schema"
+import { setupMongoConnection } from "../mongodb"
+import helmet from "helmet"
+import { isDev, updateIPDetails, isIPBlacklisted } from "../utils"
+import { WalletFactory } from "../walletFactory"
+import { gqlAdminSchema } from "../graphql"
+
+dotenv.config()
+
+const graphqlLogger = baseLogger.child({ module: "graphql" })
+
+// TODO: extract common server configs
+
+const pino_http = PinoHttp({
+  logger: graphqlLogger,
+  wrapSerializers: false,
+
+  // Define custom serializers
+  serializers: {
+    err: pino.stdSerializers.err,
+    req: pino.stdSerializers.req,
+    res: (res) => ({
+      // FIXME: kind of a hack. body should be in in req. but have not being able to do it.
+      body: res.req.body,
+      ...pino.stdSerializers.res(res),
+    }),
+  },
+  autoLogging: {
+    ignorePaths: ["/healthz"],
+  },
+})
+
+const isAuthenticated = rule({ cache: "contextual" })(async (parent, args, ctx) => {
+  if (ctx.uid === null) {
+    throw new AuthorizationError(undefined, {
+      logger: graphqlLogger,
+      request: ctx.request.body,
+    })
+  }
+  return true // ?
+})
+
+const isEditor = rule({ cache: "contextual" })(async (parent, args, ctx) => {
+  return ctx.user.role === "editor"
+})
+
+const permissions = shield(
+  {
+    Query: {
+      allLevels: and(isAuthenticated, isEditor),
+      userDetailsByPhone: and(isAuthenticated, isEditor),
+      userDetailsByUsername: and(isAuthenticated, isEditor),
+    },
+    Mutation: {
+      // currentTime: and(isAuthenticated, isEditor),
+    },
+  },
+  { allowExternalErrors: true },
+)
+
+export async function startApolloServerForAdminSchema() {
+  const app = express()
+
+  const schemaWithPermissions = applyMiddleware(gqlAdminSchema, permissions)
+
+  const server = new ApolloServer({
+    schema: schemaWithPermissions,
+    playground: process.env.NETWORK !== "mainnet",
+    introspection: process.env.NETWORK !== "mainnet",
+    context: async (context) => {
+      // @ts-expect-error: TODO
+      const token = context.req?.token ?? null
+      const uid = token?.uid ?? null
+      const ip = context.req?.headers["x-real-ip"]
+
+      if (isIPBlacklisted({ ip })) {
```

Comment:
probably the admin backend should be an "allowlist" as opposed to a "blocklist"?

and as I'm thinking through this, we probably want to use a firewall for this (or a Cloudflare of something like that) instead of trying to do it on our code directly.